//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by AsyncGenerator.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------


using System;
using NUnit.Framework;

namespace NHibernate.Test.NHSpecificTest.NH739
{
	using System.Threading.Tasks;
	using System.Threading;
	[TestFixture]
	public class FixtureAsync : BugTestCase
	{
		[Test]
		public async Task BugAsync()
		{
			int catId;

			using (ISession sess = OpenSession())
			{
				Cat c = new Cat();
				await (sess.SaveAsync(c, CancellationToken.None));
				catId = c.Id;
				await (sess.FlushAsync(CancellationToken.None));
			}

			using (ISession sess = OpenSession())
			{
				Cat c = (Cat) await (sess.GetAsync(typeof(Cat), catId, CancellationToken.None));
				Cat kitten = new Cat();
				c.Children.Add(kitten);
				kitten.Mother = c;
				await (sess.SaveAsync(kitten, CancellationToken.None));

				Assert.AreEqual(1, c.Children.Count); //Test will fail here, the c.Children.Count is 2 here

				await (sess.DeleteAsync(c, CancellationToken.None));
				await (sess.DeleteAsync(kitten, CancellationToken.None));
				await (sess.FlushAsync(CancellationToken.None));
			}
		}
	}
}